pkg/rpcmiddleware/rpcretry/stream_client.go

```diff
@@ -0,0 +1,148 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcretry
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type retriableStreamClient struct {
+	grpc.ClientStream
+	mu            sync.Mutex
+	bufferedSends []interface{}
+	streamerCall  func(ctx context.Context) (grpc.ClientStream, error)
+	counter       uint
+
+	callOpts  *options
+	parentCtx context.Context
+}
+
+func (s *retriableStreamClient) setStream(clientStream grpc.ClientStream) {
+	s.mu.Lock()
+	s.ClientStream = clientStream
+	s.mu.Unlock()
+}
+
+func (s *retriableStreamClient) getStream() grpc.ClientStream {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.ClientStream
+}
+
+func (s *retriableStreamClient) shouldRetry(err error) bool {
+	if err == nil || err == io.EOF {
+		return false
+	}
+	return isRetriable(err, s.callOpts)
+}
+
+func (s *retriableStreamClient) incrementCounter() uint {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.counter++
+	return s.counter
+}
+
+func (s *retriableStreamClient) SendMsg(m interface{}) error {
+	s.mu.Lock()
+	s.bufferedSends = append(s.bufferedSends, m)
+	s.mu.Unlock()
+	return s.getStream().SendMsg(m)
+}
+
+func (s *retriableStreamClient) CloseSend() error {
+	return s.getStream().CloseSend()
+}
+
+func (s *retriableStreamClient) Header() (metadata.MD, error) {
+	return s.getStream().Header()
+}
+
+func (s *retriableStreamClient) Trailer() metadata.MD {
+	return s.getStream().Trailer()
+}
+
+func (s *retriableStreamClient) RecvMsg(m interface{}) error {
+	err := s.getStream().RecvMsg(m)
+	if !s.shouldRetry(err) {
+		return err
+	}
+
+	for s.incrementCounter() < s.callOpts.max {
+
+		timeout := s.callOpts.timeout
+		if s.callOpts.enableMetadata {
+			header, err := s.getStream().Header()
+			if err != nil {
+				log.FromContext(s.parentCtx).Debug("Failed to retrieve metadata from stream client")
+			}
+
+			if mdTimeout := getHeaderLimiterTimeout(s.parentCtx, header, s.callOpts); mdTimeout > 0 {
+				timeout = mdTimeout
+			}
+		}
+
+		if err := waitRetryBackoff(s.parentCtx, timeout, s.counter); err != nil {
+			return err
+		}
+
+		newStream, err := s.reconnect(s.parentCtx)
+		if err != nil {
+			if isRetriable(err, s.callOpts) {
+				continue
+			}
+			return err
+		}
+
+		s.setStream(newStream)
+		err = s.getStream().RecvMsg(m)
+		if !s.shouldRetry(err) {
+			return err
+		}
+	}
+
+	return err
+}
+
+func (s *retriableStreamClient) reconnect(ctx context.Context) (grpc.ClientStream, error) {
+	s.mu.Lock()
+	bufferedSends := s.bufferedSends
+	s.mu.Unlock()
+
+	newStream, err := s.streamerCall(ctx)
+	if err != nil {
+		log.FromContext(ctx).WithError(err).Error("Failed to reconnect to server stream")
+		return nil, err
+	}
+
+	for _, msg := range bufferedSends {
```

Comment:
How can we be sure that these buffered messages weren't already processed by the server? There are so many unexpected things that can happen if we re-send already processed messages, especially since the interceptor doesn't know anything about idempotency in the application.